pkgs/by-name/ni/nixfmt-tree/package.nix

```diff
@@ -0,0 +1,133 @@
+{
+  lib,
+  runCommand,
+  buildEnv,
+  makeWrapper,
+  writers,
+  treefmt2,
+  nixfmt-rfc-style,
+  nixfmt-tree,
+
+  settings ? {
+    # The default is warn, which would be too annoying for people who just care about Nix
+    on-unmatched = "info";
+    # Assumes the user is using Git, fails if it's not
+    tree-root-file = ".git/index";
+
+    formatter.nixfmt = {
+      command = "nixfmt";
+      includes = [ "*.nix" ];
+    };
+  },
+  runtimePackages ? [
+    nixfmt-rfc-style
+  ],
+}:
+buildEnv {
+  name = "nixfmt-tree";
+
+  # Allows this derivation to be used as a shell providing both treefmt and nixfmt
+  paths = [ treefmt2 ] ++ runtimePackages;
+  pathsToLink = [ "/bin" ];
+
+  nativeBuildInputs = [
+    makeWrapper
+  ];
+
+  postBuild = ''
+    wrapProgram $out/bin/treefmt \
+      --prefix PATH : $out/bin \
+      --add-flags "--config-file ${writers.writeTOML "treefmt.toml" settings}"
+  '';
+
+  meta = {
+    mainProgram = "treefmt";
+    description = "Official Nix formatter zero-setup starter using treefmt";
+    longDescription = ''
+      A zero-setup [treefmt](https://treefmt.com/) starter to get started using the [official Nix formatter](https://github.com/NixOS/nixfmt).
+
+      - For `nix fmt` to format all Nix files, add this to the `flake.nix` outputs:
+
+        ```nix
+        formatter.''${system} = nixpkgs.legacyPackages.''${system}.nixfmt-tree;
+        ```
+
+      - The same can be done more efficiently with the `treefmt` command,
+        which you can get in `nix-shell`/`nix develop` by extending `mkShell` using
```

Comment:
Why is using `treefmt` in a shell more efficient than using `nix fmt`?

Is this just because you don't need to use flakes and therefore don't need to copy the repo to the store before eval, or is there more to it?

Maybe a full explanation isn't needed in the long description, but mentioning efficiency without an explanation peaked my curiosity 😅

Reply:
> Is this just because you don't need to use flakes and therefore don't need to copy the repo to the store before eval, or is there more to it?

Yes that, in addition to the Nix eval itself. On https://github.com/NixOS/nixpkgs/pull/380990, here are the times after changing a Nix file:
- `nix develop --command treefmt`: 9.2s
- `nix fmt`: 7.5s
- `nix-shell --run treefmt`: 5.4s
- `treefmt` (in `nix develop`/`nix-shell`): 0.2s